testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContext.cs

```diff
@@ -0,0 +1,289 @@
+using System.Linq;
+using System.Reflection;
+using MLAPI;
+using MLAPI.Messaging;
+using NUnit.Framework;
+using NUnit.Framework.Interfaces;
+using UnityEngine;
+using UnityEngine.TestTools;
+using Debug = UnityEngine.Debug;
+
+/// <summary>
+/// Allows for context based delegate execution.
+/// Can specify where you want that lambda executed (client side? server side?) and it'll automatically wait for the end
+/// of a clientRPC server side and vice versa.
+/// todo this could be used as an in-game tool too? for protocols that require a lot of back and forth?
+/// </summary>
+public class ExecuteStepInContext : CustomYieldInstruction
+{
+    public enum StepExecutionContext
+    {
+        Server,
+        Clients
+    }
+
+    [AttributeUsage(AttributeTargets.Method)]
+    public class MultiprocessContextBasedTestAttribute : NUnitAttribute, IOuterUnityTestAction
+    {
+        public IEnumerator BeforeTest(ITest test)
+        {
+            yield return new WaitUntil(() => TestCoordinator.Instance != null && hasRegistered);
+        }
+
+        public IEnumerator AfterTest(ITest test)
+        {
+            yield break;
+        }
+    }
+
+    private StepExecutionContext m_ActionContext;
+    private Action<byte[]> m_StepToExecute;
+    private string m_CurrentActionID;
+
+    // as a remote worker, I store all available actions so I can execute them when triggered from RPCs
+    public static Dictionary<string, ExecuteStepInContext> allActions = new Dictionary<string, ExecuteStepInContext>();
+    private static Dictionary<string, int> s_MethodIDCounter = new Dictionary<string, int>();
+
+    private NetworkManager m_NetworkManager;
+    private bool m_IsRegistering;
+    private List<Func<bool>> m_ClientIsFinishedChecks = new List<Func<bool>>();
+    private Func<bool> m_AdditionalIsFinishedWaiter;
+
+    private bool m_WaitMultipleUpdates;
+    private bool m_IgnoreTimeoutException;
+
+    private float m_StartTime;
+    private bool isTimingOut => Time.time - m_StartTime > TestCoordinator.maxWaitTimeout;
+    private bool ShouldExecuteLocally => (m_ActionContext == StepExecutionContext.Server && m_NetworkManager.IsServer) || (m_ActionContext == StepExecutionContext.Clients && !m_NetworkManager.IsServer);
+
+    public static bool isRegistering;
+    public static bool hasRegistered;
+    private static List<object> m_AllClientTestInstances = new List<object>(); // to keep an instance for each tests, so captured context in each step is kept
+
+    /// <summary>
+    /// This MUST be called at the beginning of each test in order to use context based steps.
+    /// Assumes this is called from same callsite as ExecuteStepInContext (and assumes this is called from IEnumerator, the method full name is unique
+    /// even with the same method name and different parameters).
+    /// This relies on the name to be unique for each generated IEnumerator state machines
+    /// </summary>
+    public static void InitContextSteps()
```

Comment:
let's keep these consistent (NetworkManager also uses Initialize, you used Initialize just a few lines below etc.)
```suggestion
    public static void InitalizeContextSteps()
```